napari_tiff/napari_tiff_metadata.py

```diff
@@ -0,0 +1,395 @@
+            contrast_limits = [contrast_limits] * shape[channel_axis]
+
+    kwargs = dict(
+        rgb=rgb,
+        channel_axis=channel_axis,
+        name=name,
+        scale=scale,
+        colormap=colormap,
+        contrast_limits=contrast_limits,
+        blending=blending,
+        visible=visible,
+    )
+    return kwargs
+
+
+def get_imagej_metadata(tif: TiffFile) -> dict[str, Any]:
+    """Return napari LayerData from ImageJ hyperstack."""
+    # TODO: ROI overlays
+    ijmeta = tif.imagej_metadata
+    series = tif.series[0]
+
+    dtype = series.dtype
+    axes = series.axes
+    shape = series.shape
+    page = series.pages[0]
+    rgb = page.photometric == 2 and shape[-1] in (3, 4)
+    mode = ijmeta.get("mode", None)
+    channels = ijmeta.get("channels", 1)
+    channel_axis = None
+
+    name = None
+    scale = None
+    colormap = None
+    contrast_limits = None
+    blending = None
+    visible = True
+
+    if mode in ("composite", "color"):
+        channel_axis = axes.find("C")
+        if channel_axis < 0:
+            channel_axis = None
+
+    if channel_axis is not None:
+        channels = shape[channel_axis]
+        channel_only = channels == ijmeta.get("images", 0)
+
+        if "LUTs" in ijmeta:
+            colormap = [Colormap(c.T / 255.0) for c in ijmeta["LUTs"]]
+        elif mode == "grayscale":
+            colormap = "gray"
+        elif channels < 8:
+            colormap = ["red", "green", "blue", "gray", "cyan", "magenta", "yellow"][
+                :channels
+            ]
+
+        if "Ranges" in ijmeta:
+            contrast_limits = numpy.array(ijmeta["Ranges"]).reshape(-1, 2)
+            contrast_limits = contrast_limits.tolist()
+
+        if channel_only and "Labels" in ijmeta:
+            name = ijmeta["Labels"]
+        elif channels > 1:
+            name = [f"Channel {i}" for i in range(channels)]
+
+        if mode in ("color", "grayscale"):
+            blending = "opaque"
+
+    elif axes[-1] == "S" and dtype == "uint16":
+        # RGB >8-bit
+        channel_axis = axes.find("S")
+        if channel_axis >= 0 and shape[channel_axis] in (3, 4):
+            rgb = False
+            n = shape[channel_axis]
+            visible = [True, True, True]
+            colormap = rgb_colormaps(samples=4)[:n]
+            name = ["Red", "Green", "Blue", "Alpha"][:n]
+            blending = ["additive", "additive", "additive", "translucent"][:n]
+        else:
+            channel_axis = None
+
+    scale = {}
+    res = page.tags.get("XResolution")
+    if res is not None:
+        scale["X"] = res.value[1] / max(res.value[0], 1)
+    res = page.tags.get("YResolution")
+    if res is not None:
+        scale["Y"] = res.value[1] / max(res.value[0], 1)
+    scale["Z"] = abs(ijmeta.get("spacing", 1.0))
+    if channel_axis is None:
+        scale = tuple(scale.get(x, 1.0) for x in axes if x != "S")
+    else:
+        scale = tuple(scale.get(x, 1.0) for x in axes if x not in "CS")
```

Comment:
If there is no scale information in metadata I strongly prefer to provide scale as `None` and allow viewer to handle it themself (based on existing layers).

Reply:
Unfortunately you cannot provide a `scale` tuple/list containing None values to napari without error:
<details>
<summary>Details (click to expand)</summary>

```python-traceback
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[24], line 1
----> 1 viewer.add_image(arr, scale=(None,None))

File ~/Documents/GitHub/napari/napari/utils/migrations.py:66, in rename_argument.<locals>._wrapper.<locals>._update_from_dict(*args, **kwargs)
     64     kwargs = kwargs.copy()
     65     kwargs[to_name] = kwargs.pop(from_name)
---> 66 return func(*args, **kwargs)

File ~/Documents/GitHub/napari/napari/components/viewer_model.py:947, in ViewerModel.add_image(self, data, channel_axis, affine, attenuation, blending, cache, colormap, contrast_limits, custom_interpolation_kernel_2d, depiction, experimental_clipping_planes, gamma, interpolation2d, interpolation3d, iso_threshold, metadata, multiscale, name, opacity, plane, projection_mode, rendering, rgb, rotate, scale, shear, translate, visible)
    939     if k not in iterable_kwargs and is_sequence(v):
    940         raise TypeError(
    941             trans._(
    942                 "Received sequence for argument '{argument}', did you mean to specify a 'channel_axis'? ",
   (...)
    945             )
    946         )
--> 947 layer = Image(data, **kwargs)
    948 self.layers.append(layer)
    950 return layer

File ~/Documents/GitHub/napari/napari/layers/base/base.py:111, in PostInit.__call__(self, *args, **kwargs)
    110 def __call__(self, *args, **kwargs):
--> 111     obj = super().__call__(*args, **kwargs)
    112     obj._post_init()
    113     return obj

File ~/Documents/GitHub/napari/napari/utils/migrations.py:66, in rename_argument.<locals>._wrapper.<locals>._update_from_dict(*args, **kwargs)
     64     kwargs = kwargs.copy()
     65     kwargs[to_name] = kwargs.pop(from_name)
---> 66 return func(*args, **kwargs)

File ~/Documents/GitHub/napari/napari/layers/image/image.py:266, in Image.__init__(self, data, affine, attenuation, blending, cache, colormap, contrast_limits, custom_interpolation_kernel_2d, depiction, experimental_clipping_planes, gamma, interpolation2d, interpolation3d, iso_threshold, metadata, multiscale, name, opacity, plane, projection_mode, rendering, rgb, rotate, scale, shear, translate, visible)
    263     rgb = rgb_guess
    265 self.rgb = rgb
--> 266 super().__init__(
    267     data,
    268     affine=affine,
    269     blending=blending,
    270     cache=cache,
    271     custom_interpolation_kernel_2d=custom_interpolation_kernel_2d,
    272     depiction=depiction,
    273     experimental_clipping_planes=experimental_clipping_planes,
    274     metadata=metadata,
    275     multiscale=multiscale,
    276     name=name,
    277     ndim=len(data_shape) - 1 if rgb else len(data_shape),
    278     opacity=opacity,
    279     plane=plane,
    280     projection_mode=projection_mode,
    281     rendering=rendering,
    282     rotate=rotate,
    283     scale=scale,
    284     shear=shear,
    285     translate=translate,
    286     visible=visible,
    287 )
    289 self.rgb = rgb
    290 self._colormap = ensure_colormap(colormap)

File ~/Documents/GitHub/napari/napari/layers/intensity_mixin.py:31, in IntensityVisualizationMixin.__init__(self, *args, **kwargs)
     30 def __init__(self, *args, **kwargs) -> None:
---> 31     super().__init__(*args, **kwargs)
     33     self.events.add(
     34         contrast_limits=Event,
     35         contrast_limits_range=Event,
     36         gamma=Event,
     37         colormap=Event,
     38     )
     39     self._gamma = 1

File ~/Documents/GitHub/napari/napari/layers/_scalar_field/scalar_field.py:212, in ScalarFieldBase.__init__(self, data, affine, blending, cache, custom_interpolation_kernel_2d, depiction, experimental_clipping_planes, metadata, multiscale, name, ndim, opacity, plane, projection_mode, rendering, rotate, scale, shear, translate, visible)
    209 if ndim is None:
    210     ndim = len(data.shape)
--> 212 super().__init__(
    213     data,
    214     ndim,
    215     name=name,
    216     metadata=metadata,
    217     scale=scale,
    218     translate=translate,
    219     rotate=rotate,
    220     shear=shear,
    221     affine=affine,
    222     opacity=opacity,
    223     blending=blending,
    224     visible=visible,
    225     multiscale=multiscale,
    226     cache=cache,
    227     experimental_clipping_planes=experimental_clipping_planes,
    228     projection_mode=projection_mode,
    229 )
    231 self.events.add(
    232     attenuation=Event,
    233     custom_interpolation_kernel_2d=Event,
   (...)
    246     rendering=Event,
    247 )
    249 self._array_like = True

File ~/Documents/GitHub/napari/napari/layers/base/base.py:334, in Layer.__init__(self, data, ndim, affine, blending, cache, experimental_clipping_planes, metadata, mode, multiscale, name, opacity, projection_mode, rotate, scale, shear, translate, visible)
    331     name = magic_name(data)
    333 if scale is not None and not np.all(scale):
--> 334     raise ValueError(
    335         trans._(
    336             "Layer {name} is invalid because it has scale values of 0. The layer's scale is currently {scale}",
    337             deferred=True,
    338             name=repr(name),
    339             scale=repr(scale),
    340         )
    341     )
    343 # Needs to be imported here to avoid circular import in _source
    344 from napari.layers._source import current_source

ValueError: Layer 'arr' is invalid because it has scale values of 0. The layer's scale is currently (None, None)
```

</details>

If *all* the scale values are None, then you can do `viewer.add_image(data, scale=None)` without error.

But if only some of the scale values for certain axes are available, then this causes an error. 

I could choose to use an obviously nonsense scale dummy value, like -1 for the missing axes ...but that would then be inconsistent with napari's internal behaviour. While napari will accept `viewer.add_image(data, scale=None)`, when you ask the layer what scale value it has, napari will return `viewer.layers[-1].scale = [1.0, 1.0]`. 

I think it would be unwise to create an inconsistency in behaviour.

So while I do think your complaint about default scale values is valid, it belongs better as an issue in the napari repository.

Reply:
Issue created here for you, @Czaki  https://github.com/napari/napari/issues/6968